modules/lang-painless/src/main/java/org/elasticsearch/painless/MethodWriter.java

```diff
@@ -262,6 +262,50 @@ public void writeToStrings() {
         }
     }
 
+    /** Writes a dynamic binary instruction: returnType, lhs, and rhs can be different */
+    public void writeDynamicBinaryInstruction(Location location, Type returnType, Type lhs, Type rhs, Operation operation) {
+        org.objectweb.asm.Type descriptor = org.objectweb.asm.Type.getMethodType(returnType.type, lhs.type, rhs.type);
```

Comment:
I know, i am just exhausted from the slave labor on this one already. I did try to do the refactorings you suggested, to try to make it cleaner and also remove string constants and stuff. But it got messy quickly (the root problem, Operator is syntax oriented).

I'm not gonna leave it alone. I just want to iterate here. There is a lot to followup with. For example return values of unary operators, compound assignments, all other cases where we do inefficient casts.